src/command/shim.rs

```diff
@@ -100,38 +108,41 @@ Options:
     }
 }
 
-fn list(session: &mut Session, verbose: bool) -> Fallible<bool> {
+fn list(session: &Session, verbose: bool) -> Fallible<bool> {
     path::shim_dir()
         .and_then(|shim_dir| fs::read_dir(shim_dir).unknown())
         .map(|files| {
-            files
-                .map(|file| {
-                    file.and_then(|f| {
-                        f.path().file_name().map_or(Ok(false), |shim_name| {
-                            if verbose {
-                                match resolve_shim(session, &shim_name) {
-                                    Ok(shim_info) => {
-                                        println!("{} -> {}", shim_name.to_string_lossy(), shim_info)
-                                    },
-                                    Err(err) => {
-                                        style::display_error(style::ErrorContext::Notion, &err);
-                                        return Ok(false);
-                                    },
-                                }
-                            } else {
-                                println!("{}", shim_name.to_string_lossy());
-                            }
-                            Ok(true)
-                        })
-                    })
-                })
-                .collect::<Vec<_>>()
-                .iter()
-                // return false if anything failed
-                .all(|ref result| result.as_ref().ok() == Some(&true))
+            files.map(|file| {
+                file.and_then(|f| print_file_info(f, session, verbose))
+            })
+            .collect::<Vec<_>>()
+            .iter()
+            // return false if anything failed
+            .all(|ref result| result.as_ref().ok() == Some(&true))
         })
 }
 
+fn print_file_info(
+    file: fs::DirEntry,
+    session: &Session,
+    verbose: bool,
+) -> Result<bool, io::Error> {
+    file.path().file_name().map_or(Ok(false), |shim_name| {
```

Comment:
A `DirEntry` always has a file name, so instead of using the `Option` that comes from `file.path().file_name()` you can use `file.file_name()` which directly returns an `OsString`.

Reply:
Doh! I missed that in the docs somehow